File: 0386-lexicographical-numbers/0386-lexicographical-numbers.go
func lexicalOrder(n int) []int {
       ans := make([]int, 0, n)

    var f func(i int)
    f = func(i int) {
        if i > n {
            return
        }
        if i > 0 {
            ans = append(ans, i)
        }
        for j := 0; j <= 9; j++ {
            if i == 0 && j == 0 {
                continue
            }
            x := 10*i + j
            if x > n {
                break
            }
            f(x)
        }
    }

    f(0)
    return ans
}

File: 0386-lexicographical-numbers/0386-lexicographical-numbers.py
class Solution:
    def __init__(self):
        self.ans = []
        self.n = 0

    def f(self, i):
        if i > self.n:
            return
        if i > 0:
            self.ans.append(i)
        for j in range(1 if i == 0 else 0, 10):
            x = 10 * i + j
            if x > self.n:
                break
            self.f(x)

    def lexicalOrder(self, n):
        self.n = n
        self.ans = []
        self.f(0)
        return self.ans
